Created a parameter and some logic to allow EC2 machines to continue running for reuse

joshuafarrell opened the pull request:
We had a setup where tests were not running for a full hour, and rather than be charged for every instance created we wanted to reuse the machines for the full time.

Added a parameter called ec2.keep.alive that can be set true/false when set to true it will override ec2.force.termination and allow the EC2 instances to continue running.

Along with this I added some code at the beginning of AwsGatlingRunner.launchEC2Instances() function that searches for any running instances with the GATLING_NAME_TAG, and if it finds any it will load up the instances variable with them so they can be reused and skip the creation step.

These steps work best when you run install_gatling.sh and then make a template machine from which to boot new EC2s.

Commit:
Added output for existing instances

First steps to be able to reuse an instance within the hour window because we get charged for it anyways.

Commit:
Initial check for running EC2 instance

Finished implementing the first part of reusing EC2 instances, still need to implement a parameter to not terminate machine at the end of a short run, then we can reuse for the whole hour we are getting charged for.

Commit:
Remove code from master.

Just cleaning up the master branch until the other changes are completed.

Commit:
Add conditional for re-using EC2 machine

Checks for existing EC2 machine with specific filter and if none are found we spin up a new one.

Commit:
Add conditional for re-using EC2 machine

Finished implementing a parameter to keep an EC2 machine running, will override ec2.force.termination.

Commit:
Merge remote-tracking branch 'remotes/origin/master' into changes

Conflicts:
	.idea/compiler.xml
	.idea/gatling-aws-maven-plugin.iml
	.idea/misc.xml
	src/main/java/com/ea/gatling/AwsGatlingRunner.java

Commit:
Merge remote-tracking branch 'remotes/origin/master' into changes

Conflicts:
	.idea/compiler.xml
	.idea/gatling-aws-maven-plugin.iml
	.idea/misc.xml
	src/main/java/com/ea/gatling/AwsGatlingRunner.java

ingojaeckel commented on src/main/java/com/ea/gatling/AwsGatlingRunner.java:
```diff
@@ -37,52 +37,72 @@ public AwsGatlingRunner() {
+        filters[0] = new Filter("tag:Name").withValues(GATLING_NAME_TAG.getValue());
+        filters[1] = new Filter("instance-state-name").withValues("running");
+
+        DescribeInstancesResult describeInstancesResult = ec2client.describeInstances(new DescribeInstancesRequest().withFilters(filters));
```
Can you bring back the old multi-line formatting please?

ingojaeckel commented on src/main/java/com/ea/gatling/AwsGatlingRunner.java:
```diff
@@ -37,52 +37,72 @@ public AwsGatlingRunner() {
+        // Check for existing EC2 instances that fit the filter criteria and use those
+        for (Reservation reservation : describeInstancesResult.getReservations()) {
+            for (Instance instance : reservation.getInstances()) {
+                System.out.println("Reservations " + instance.getInstanceId() + " (" + instance.getState().getName() + "): " + instance.getSecurityGroups().get(0).getGroupName());
```
This gets pretty hard to read. Do you mind changing it to a System.out.format() call instead? See line 67 for an example.

ingojaeckel commented on src/main/java/com/ea/gatling/AwsGatlingRunner.java:
```diff
@@ -37,52 +37,72 @@ public AwsGatlingRunner() {
+                    }
+                }
+
+                DescribeInstanceStatusRequest describeInstanceStatusRequest = new DescribeInstanceStatusRequest().withInstanceIds(instances.keySet());
```
do you mind adding a few comments to the new code?

ingojaeckel commented on src/main/java/com/ea/gatling/GatlingAwsMojo.java:
```diff
@@ -54,6 +54,9 @@
     private boolean ec2ForceTermination = false;
 
+    @Parameter(property = "ec2.keep.alive", defaultValue="false")
+    private boolean ec2KeepAlive = false;
```
Do you mind adding a comment to the new parameter? Please also document it in the README.

ingojaeckel commented on .idea/compiler.xml:
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
please remove all .idea/\* files from the pull request.

ingojaeckel commented:
@joshuafarrell thanks for this pull request. I think this is a great idea. Let me know what you think about some of my comments.

joshuafarrell committed:
Removed .idea folder from repo

joshuafarrell committed:
Removed the .idea directory, added comments, and multi-lined line 47.

joshuafarrell commented:
Thanks for the awesome plugin it saved me a lot of time getting my framework ready for outside testing and I did not have to go back to Jmeter/Blazemeter.

I've attempted fixing the changes and adding more to the comments.

Still need to update the readme though.

joshuafarrell committed:
Update README.md

Added ec2.keep.alive to the maven properties table with a description and example of when to use.

joshuafarrell committed:
Added another filter for existing instances and one more output.

The output states that no existing instances was found and it is creating a new one, and the filter is for the instance-type, so we can mix and match tests of different sizes and not have a conflict. (Just in case)

ingojaeckel commented on README.md:
```diff
@@ -231,6 +231,7 @@ This will spin up 3 c3.large instances and start the com.FooTest simulation on e
 | ec2.key.pair.name | gatling-key-pair | The [key pair](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-key-pairs.html) that will be used to access the load generator instances. |  | 1.0.0 |
 | ec2.security.group | gatling-security-group | The [security group](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-network-security.html) used by the load generators. Make sure that your local development environment and CI system have SSH access to the load generators via the security group.  |  | 1.0.0 |
 | ec2.force.termination | false | `true` if the load generators should be terminated after the test despite of test failures. `false` if the load generators should be kept alive in case of failures to allow troubleshooting. If the tests succeeds all load generators will be terminated regardless of this setting. Setting it to `false` allows troubleshooting failed load generators. Setting it to `tr ...
+| ec2.keep.alive | false | `true` if the load generators should be left running for another test (will override ec2.force.termination. `false` to allow load generators to be terminated. | If tests are not going to run for the full hour use this setting to keep from being charged for the 1 hour minimum up time, also cuts down on load generator initializing time | |
```
insert the missing ) after "will override ..".

ingojaeckel replied:
please mention that this is available since version 1.0.4-SNAPSHOT. use the rightmost column for that. i will change to 1.0.4 after releasing the new version.

ingojaeckel commented:
@joshuafarrell looks pretty good. I will accept the PR once you addressed my concern about the readme. Then I will retest this and release the new version.

joshuafarrell committed:
Modified the README.md

Added missing parenthesis, and the available version to 1.0.4-SNAPSHOT

joshuafarrell commented:
@ingojaeckel I noticed that when reusing machines the reports will begin to stack and are all combine, resulting in some weird reporting.  I added a line to the install-gatling.sh file that attempts to clear the items in the results folder "rm -r gatling-charts-highcharts-bundle-2.2.0-M3/results/*"
